### ISSUE TITLE
aws_cloudwatch_log_metric_filter: default_value is automatically set to 0

### ISSUE THREAD

**Issue by iamam34** [terraform-providers/terraform-provider-aws#2384]
When I apply a terraform configuration that does NOT include default_value, terraform creates a log filter in AWS that has the default_value set to 0.

### Terraform Version
Terraform v0.10.3 with aws provider v1.3.1

### Affected Resource(s)
- aws_cloudwatch_log_metric_filter

### Terraform Configuration Files
```hcl
resource "aws_cloudwatch_log_metric_filter" "my-function-memory-used" {
  name           = "MyFunction_MemoryUsed"
  pattern        = "[type=REPORT, ... , memory_used, unit]"
  log_group_name = "/aws/lambda/MyFunction"

  metric_transformation {
    name      = "MemoryUsed"
    namespace = "MyFunction.CustomMetrics"
    value     = "$memory_used"
  }
}
```

### Expected Behavior
Creates a log filter with default_value **not set**, so that log lines that don't match the filter don't output any value for the metric. http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/MonitoringLogData.html

### Actual Behavior
Creates a log filter with default_value **set to 0**, so that every log line outputs a value for the metric (non-matching lines output 0). 

### Steps to Reproduce
1. Write terraform for a log_filter WITHOUT a default_value
1. `terraform apply` (note that the console output does not mention default_value at all)
    ```
    aws_cloudwatch_log_metric_filter.my-function-memory-used: Creating...
    [11:12:27][Step 2/2]   log_group_name:                    "" => "/aws/lambda/MyFunction"
    [11:12:27][Step 2/2]   metric_transformation.#:           "" => "1"
    [11:12:27][Step 2/2]   metric_transformation.0.name:      "" => "MemoryUsed"
    [11:12:27][Step 2/2]   metric_transformation.0.namespace: "" => "MyFunction.CustomMetrics"
    [11:12:27][Step 2/2]   metric_transformation.0.value:     "" => "$memory_used"
    [11:12:27][Step 2/2]   name:                              "" => "MyFunction_MemoryUsed"
    [11:12:27][Step 2/2]   pattern:                           "" => "[type=REPORT, ... , memory_used, unit]"
    ```
1. Visit the AWS Console CloudWatch and see that your new metric has a minimum value of 0 and an unexpectedly low average. 

### References
- #1578 - this PR from a month ago added support for the optional default_value

**Comment by iamam34** [comment]
@stack72 @Ninir Since you were involved in the earlier work on this attribute, do you have any special insights?

**Comment by Ninir** [comment]
Hi @iamam34 

I think I've found the issue (which should be easy to resolve by the way): this attribute is of type Float, which seem to imply that the value is always provided as 0 when not set... as the check deciding whether the value should be set in the API call is a check against `""`, the value is always sent... and as 0 is a valid value... I think we should just go with having a default to -1 or updating the type from float to string.

If someone wants to take over this, just go :) otherwise will fix it later :)

**Comment by iamam34** [comment]
Great to hear it's straightforward to fix! 
Note: If set, the value must be a float. If not set, we need to _not send_ the value (not even -1... but null might be ok; I haven't checked the AWS API docs). 

Up to you whether a empty string is considered valid or invalid terraform syntax (but since it's not a float, we still don't want to send it).

**Comment by jocelynthode** [comment]
Hey I'm trying to work on this issue. As it is my first time working on Terraform and I'm really new to Go, I'll probably need guidance. I'll open a WIP PR shortly.

**Comment by FrederikNygaardSvendsen** [comment]
Any news? This has unknowingly rendered our 'ResponseTime' metrics useless (since they are cluttered with zeros)

**Comment by jocelynthode** [comment]
I asked for help on my PR as I'm rather new to this however nobody answered.

**Comment by skynardo** [comment]
I also have this issue. It seems that if I manually edit the Log Metric Filter via AWS Console and delete the zero, the Default Value field changes to "none". This fixes the issue as long as no changes are made to the resource via terraform (aws provider 1.6).

**Comment by zytek** [comment]
Just got hit by this when creating alarm on `SampleCount >=0`. As there are two almost identical PRs open to address it please advise how we could move forward with fixing this issue.

**Comment by MooreDerek** [comment]
Have encountered this one as well. Would be good to know if a fix is on the way. The "-1" seems to be a good solution.

**Comment by bflad** [comment]
The fix for this should be merged and will release with version 1.44.0 of the AWS provider, likely tomorrow. Apologies for the long delays here.

**Comment by bflad** [comment]
This has been released in [version 1.44.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1440-november-14-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
